Bottom: 1003240083c65937babdef458657314f0351ca64
Top:    13c148f833b9e2e711e487c36233fc853cbe5b98
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-01-06 21:11:39 +0000

Fixup some bugs from the added soldermask layers

(TODO: Finish this fixup and merge it down into the relevant patches)


---

diff --git a/src/find.c b/src/find.c
index 3bd66f6..4edb048 100644
--- a/src/find.c
+++ b/src/find.c
@@ -776,6 +776,8 @@ LookupLOConnectionsToLOList (int flag, bool AndRats)
   Cardinal i, group, layer, ratposition,
     lineposition[MAX_LAYER],
     polyposition[MAX_LAYER], arcposition[MAX_LAYER], padposition[2];
+  Cardinal top_group = GetLayerGroupNumberBySide (TOP_SIDE);
+  Cardinal bottom_group = GetLayerGroupNumberBySide (BOTTOM_SIDE);
 
   /* copy the current LO list positions; the original data is changed
    * by 'LookupPVConnectionsToLOList()' which has to check the same
@@ -848,22 +850,18 @@ LookupLOConnectionsToLOList (int flag, bool AndRats)
                         (POLYGONLIST_ENTRY (layer, *position), group, flag, AndRats))
                       return (true);
                 }
-              else
-                {
-                  /* try all new pads */
-                  layer -= max_copper_layer;
-                  if (layer > 1)
-                    {
-                      Message (_("bad layer number %d max_copper_layer=%d in find.c\n"),
-                               layer, max_copper_layer);
-                      return false;
-                    }
-                  position = &padposition[layer];
-                  for (; *position < PadList[layer].Number; (*position)++)
-                    if (LookupLOConnectionsToPad
-                        (PADLIST_ENTRY (layer, *position), group, flag, AndRats))
-                      return (true);
-                }
+            }
+
+          /* try all new pads */
+          if (group == top_group || group == bottom_group)
+            {
+              int side = (group == top_group) ? TOP_SIDE : BOTTOM_SIDE;
+
+              position = &padposition[side];
+              for (; *position < PadList[side].Number; (*position)++)
+                if (LookupLOConnectionsToPad
+                    (PADLIST_ENTRY (side, *position), group, flag, AndRats))
+                  return (true);
             }
         }
 
diff --git a/src/hid/gtk/gui-config.c b/src/hid/gtk/gui-config.c
index eb8fbd1..0c48026 100644
--- a/src/hid/gtk/gui-config.c
+++ b/src/hid/gtk/gui-config.c
@@ -1513,6 +1513,15 @@ config_layer_group_button_state_update (void)
   for (g = 0; g < max_group; g++)
     for (i = 0; i < layer_groups.Number[g]; i++)
       {
+	/* Skip anything beyond the bottom silk / top silk layers residing
+	   at index max_copper_layer + 0 and max_copper_layer + 1. These are
+	   used to indicate the groups for bottom and top sides respectively,
+	   so form somewhat of a special case in our handling here (they have
+	   group buttons).
+	 */
+	if (layer_groups.Entries[g][i] >= max_copper_layer + 2)
+	  continue;
+
 /*			printf("layer %d in group %d\n", layer_groups.Entries[g][i], g +1); */
 	config_layer_group[layer_groups.Entries[g][i]] = g + 1;
 	gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON
